wsmux/stream.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	return nil
+}
+
+// SetWriteDeadline sets the write timer
+func (s *stream) SetWriteDeadline(t time.Time) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	//stop timer if not nil
+	if s.writeTimer != nil {
+		_ = s.writeTimer.Stop()
+		s.writeTimer = nil
+	}
+	// clear deadline exceeded
+	s.writeDeadlineExceeded = false
+	if !t.IsZero() {
+		delay := t.Sub(time.Now())
+		s.writeTimer = time.AfterFunc(delay, s.onExpired(&s.writeDeadlineExceeded))
+	}
+
+	return nil
+}
+
+// UnblockAndBroadcast unblocks bytes and broadcasts so that writes can
+// continue
+func (s *stream) UnblockAndBroadcast(read uint32) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	defer s.session.logger.Printf("unblock broadcasted : stream %d", s.id)
+	s.unblocked += read
+}
+
+// PushAndBroadcast adds data to the read buffer and broadcasts so that
+// reads can continue
+func (s *stream) PushAndBroadcast(buf []byte) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	defer s.session.logger.Printf("push broadcasted : stream %d", s.id)
+	_, err := s.b.Write(buf)
+	s.endErr = err
+}
+
+// AcceptStream accepts the current stream by closing the accepted channel
+func (s *stream) AcceptStream(read uint32) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	s.unblocked += read
+	s.state = accepted
+	close(s.accepted)
+
+}
+
+// SetDeadline sets the read and write deadlines for the stream
+func (s *stream) SetDeadline(t time.Time) error {
+	if err := s.SetReadDeadline(t); err != nil {
+		s.endErr = err
+		return err
+	}
+	if err := s.SetWriteDeadline(t); err != nil {
+		s.endErr = err
+		return err
+	}
+	return nil
+}
+
+func (s *stream) IsRemovable() bool {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.state == dead && s.b.Len() == 0
+}
+
+// setRemoteClosed sets the value of stream.remoteClosed. This indicates that the remote has sent a fin packet
+func (s *stream) setRemoteClosed() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	if s.state == closed {
+		s.state = dead
+	} else {
+		s.state = remoteClosed
+	}
+}
+
+//LocalAddr returns the local address of the underlying connection
+func (s *stream) LocalAddr() net.Addr {
+	return s.session.conn.LocalAddr()
+}
+
+// RemoteAddr returns the remote address of the underlying connection
+func (s *stream) RemoteAddr() net.Addr {
+	return s.session.conn.RemoteAddr()
+}
+
+// Close closes the stream
+func (s *stream) Close() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+
+	switch s.state {
+	// return nil if already closed
+	case closed:
+		return nil
+	case remoteClosed:
+		s.state = dead
+	default:
+		s.state = closed
+	}
+
+	if err := s.session.send(newFinFrame(s.id)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read reads bytes from the stream
+func (s *stream) Read(buf []byte) (int, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	s.session.logger.Printf("stream %d: read requested", s.id)
+
+	for s.b.Len() == 0 && s.endErr == nil {
```

Comment:
You should put all the conditions that makes the wait stop into the whole condition... And have the body be only c.Wait()

The deal with each condition afterwards... In if statements...

Well, maybe I'm just thinking it might be easier that way... As the whole loop is only used for waiting... And what happens after is separate..